working/0546-patterns/patterns-feature-specification.md

```diff
@@ -1227,69 +1347,90 @@ var (a, int b) = ... // Schema is `(?, int)`.
 
 The context type schema for a pattern `p` is:
 
-*   **List binder or matcher**: A type schema `List<E>` where:
-    *   If `p` has a type argument, then `E` is the type argument.
-    *   Else `E` is the greatest lower bound of the type schemas of all element
-        subpatterns.
-
-*   **Map binder or matcher**: A type schema `Map<K, V>` where:
-    *   If `p` has type arguments then `K`, and `V` are those type arguments.
-    *   Else `K` is the least upper bound of the types of all key expressions
-        and `V` is the greatest lower bound of the context type schemas of all
-        value subpatterns.
+*   **Or**: The least upper bound of the context type schemas of the branches.
 
-*   **Record binder or matcher**: A record type scheme with positional and named
-    fields corresponding to the type schemas of the corresponding field
-    subpatterns. If a named field uses the shorthand syntax to infer a variable
-    subpattern with the same name as the field, then the type schema is `?` for
-    that field.
+*   **And**: The greatest lower bound of the context type schemas of the branches.
 
-    *Note that the type schema will be a record type even when the matched value
-    type isn't a record, as in:*
+*   **Null-check** or **null-assert**: A context type schema `E?` where `E` is
+    the context type schema of the inner pattern. *For example:*
 
     ```dart
-    var p = Point(x: 1, y: 2);
-    var (x: x, y: y) = p;
+    var [[int x]!] = [[]]; // Infers List<List<int>?> for the list literal.
     ```
 
-    *The record type schema inferred from the pattern may be used to infer
-    record field types on the initialized value if the value is a record
-    literal, and may otherwise be captured. But if the initializer type isn't a
-    record, then the record type context schema inferred from the pattern ends
-    up being ignored.*
+*   **Literal** or **constant**: The context type schema is the static type of
+    the pattern's constant value expression.
 
-*   **Variable matcher**:
-    *   If `p` has a type annotation, the context type schema is `Object?`.
-        *It is not the annotated type because a variable matching pattern can
-        be used to downcast from any other type.*
-    *   Else it is `?`.
+*   **Variable**:
 
-*   **Cast binder**, **variable binder**, **wildcard matcher**, or **extractor
-    matcher**: The context type schema is `Object?`.
+    *   If the pattern is in a refutable context, the context type schema is
+        `Object?`. *We ignore the type annotation in a refutable context because
```

Comment:
We do now because I just tweaked it to allow context types in pattern-if. But I also removed these special cases because if we're going to use a context type there, we may as well use a precise one. That way, a pattern in a pattern-if that *could* be just an irrefutable destructuring can be if downwards inference makes it such:

```dart
List<T> makeList<T>() => [];

if ([int i] = makeList()) { ... }
```

Using `Object?` as the context type schema for `int i` would lead the `[int i]` pattern to have context type schema `List<Object>`. That would in turn lead downwards inference to infer `makeList<Object>`. Which means now the `[int i]` pattern would have to be refutable to check that the returned list is actually a `List<int>`.

But there's no need for that to be refutable at all if we use `int` as the context type schema for `int i`. So now it unconditionally uses a precise context type schema regardless of whether its in a refutable context.